### DISCUSSION TITLE
Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 8 bytes (0 received)

### DISCUSSION THREAD

**Discussion by u4h37u** [nathanmarlor/foxess_modbus#676] [Support] [answered]
I'm struggling to get Home Assistant to communicate with the inverter using this excellent looking integration.

Inverter: H1-6.0-E-G2 with the 16 pin connector
Inverter Hardware Version : -Master : 1.27Slave : 1.02Manager : 1.44
Waveshare RS485 to ETH (B) running v1.490
HomeAssistant Core 2024.8.3 Supervisor 2024.09.1

Port and slave IDs are set to defaults and checked on the inverter.

The wires had been connected to the modbus by my installer, but I took it apart to check.  They had been put into the wrong pins.  I'd also been getting some CT clamp errors, and you can see why in the "before" photo, now corrected in the "after" photo.  I checked three times that there was a good connection and in the correct pins.

![IMG_20240913_132614](https://github.com/user-attachments/assets/43b59209-7b69-4313-8bbd-6871d645b589)
![IMG_20240913_143920](https://github.com/user-attachments/assets/507021c1-8bf1-40a3-8085-ba70b18eeeb6)

I'd really hoped having done this that it would all be good.

When I try to connect to the modbus using the HA integration setup I do see the waveshare link light flash blue, so HA is communicating with the waveshare fine.  

I've also re-checked all of the settings on the waveshare against the images posted on the very clear guide at:

https://github.com/nathanmarlor/foxess_modbus/wiki/Waveshare-RS485-to-ETH-%28B%29

![Screenshot 2024-09-13 160544](https://github.com/user-attachments/assets/68ad2514-3168-45d8-9008-8b75c10a3863)
![Screenshot 2024-09-13 160702](https://github.com/user-attachments/assets/18e48eb0-6830-4e93-b3a4-499d4d931a0c)

I've tried multi-host on and off (mentioned as optional but not needed in my setup).

The closest post to this is: https://github.com/nathanmarlor/foxess_modbus/discussions/610 - which was solved by switching to a twisted pair cable, which I'm using (though only one wire in each pair is used, assume that's the usual way people would wire it).   I can see that they turned on pymodbus logging, but I can't find instructions for how to do that.

I've spent several hours tying to suss this, nothing in other posts seems to quite match; anything obvious I'm doing wrong or any ideas of things to try?  

Thank you 🙏

**Comment by FozzieUK** [comment]
It looks setup correctly, and the wiring is much tidier now :)

I can’t see the pin numbers but you should be connected to pins 3 & 4 for RS485B & A

You only need to worry about twisted pair and shielding if it’s going long distance so a single pair is fine.

Your inverter firmware versions are good as well so that’s another thing ticked.

You should see the ACT light flash when you address the waveshare, if it’s lit permanently it usually means you have the A & B wires swapped.

So a couple of things to check on the inverter, make sure in settings that the ethernet DHCP is disabled this sometimes overrides the RS485 port and defaults comms to ethernet, and also that the slave is set to 247 (sometimes the inverter will restart at this point)

If you could try connecting the integration again and copy any logs that relate to it below please.

**Comment by u4h37u** [comment]
Hello Dave - thank you for replying so quickly to my post.  Thank you also for commenting on the wiring improvement :-)

I have a new G2 model which doesn't seem to have an ethernet port and only RS485 is available in the communications menu.  Slave is set to 247.

I don't ever see the ACT light on the waveshare....  

I powered down the inverter this morning and used a multimeter to test the connection from the waveshare to the port output going into the inverter and they're connected.  I'm also convinced that I have pins 3 & 4 connected the right way round, though I've also tried switching them over this morning just through exasperation but no success.

HA logs report:

```
2024-09-14 10:53:37.814 ERROR (MainThread) [custom_components.foxess_modbus.modbus_controller] Autodetect: failed to connect to (tcp://192.168.4.73:502)
pymodbus.exceptions.ModbusIOException: Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 8 bytes (0 received)
File "/config/custom_components/foxess_modbus/modbus_controller.py", line 502, in autodetect
File "/config/custom_components/foxess_modbus/client/modbus_client.py", line 134, in read_registers
raise ModbusClientFailedError(message, self, response) from response
custom_components.foxess_modbus.client.modbus_client.ModbusClientFailedError: Error reading registers. Type: RegisterType.HOLDING; start: 30000; count: 15; slave: 247 from tcp://192.168.4.73:502: Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 8 bytes (0 received)
```

I've ordered one of these this morning to test the waveshare device is actually working - at least one of the many forum posts I've read suggested that their waveshare device was duff:

https://www.ebay.co.uk/itm/264814575926

Frustratingly also this morning the inverter has decided to ignore the battery, drawing from grid to make up any excess requirement from load.  CT clamp data and load data seem accurate, min soc 10%, min soc on grid 15% (I use EPS so this could probably do with going up a bit), max soc 100%, nothing set in schedulers (never even used them), all 0's in the time on battery force charge settings, but this is another story/new issue.  🤦

**Comment by FozzieUK** [comment]
I’ve always found the waveshare very reliable but as you say a few users have had initial setup problems, which were mostly cured by the ‘full reset’ shown on the wiki page and then set it up again, after which it’s worked.

I saw your post over on the community website about the battery not working, i’ve answered that over there - hopefully one of those points will get it back up running.

**Reply by u4h37u** [reply]
Thanks Dave, you're a star for the help you offer - do you have a "buy me a coffee" link or similar?  Thanks also for mentioning the post on the community site, it doesn't seem to notify of follow ups to posts unless you subscribe so I didn't see your response straight away.

It will take me a while to do the next steps of a full reset and sniff the serial port coming out to test the waveshare isn't faulty.  In many ways I really hope it is as I don't fancy my chances of convincing FoxESS that the inverter modbus pins 3 & 4 are not working as that would be the next step.

Wish things were simple for once!  

At least once I've sussed this, which I will, I'll know quite a lot to help others as well :-)

**Reply by FozzieUK** [reply]
I’ve only seen one modbus RS485 port failure (after a power cut) and I was surprised how quickly Fox diagnosed it, there’s supposed to be a voltage across the pins which was absent so the inverter was swapped out. I’m hoping yours is a much simpler config problem.

I’ve replied on the community website but as you say sadly it doesn’t send notfications - It feels like the CT is at the heart of this problem, if the inverter cannot correctly measure the load it won’t be able to balance it with the batteries - so it is possible that this is the problem. 
The Energy Flow on the fox cloud site is usually quite good at diagnosing that - is there a load being measured at all (is it negative) etc.. - as you said you had random problems in the past i’m wondering if there is an actual cable or CT clamp problem, but you’ll need your installer or Fox to diagnose that.

**Reply by u4h37u** [reply, marked as answer]
Solution found.

The original issue was the modbus had been incorrectly wired by the installer which I fixed, but still wasn't working.

The second thing needed was a replacement ethernet cable into the Waveshare RS485 device.  The original cable was new, and I happen to own a cable tester so when troubleshooting the original issue over many hours I'd tested the cable which was apparently fine. Out of desperation and having tried everything else I could think of I tried a change of ethernet cable to another identical cat6 cable, and suddenly it works.  HA connected and all the metrics are coming through.

Thanks for all your help @FozzieUK, I'm very very happy with it all now!
